From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: gennro <gennro2004@gmail.com>
Date: Tue, 6 Oct 2026 19:47:33 -0500
Subject: [PATCH] hexagonrpcd: serve the sensors registry version file

The sensors process reads /mnt/vendor/persist/sensors/registry/sns_reg_version
next to the registry. Without it, it tries to write a new one (and a new
registry), which hexagonrpcd can't serve. Map it from sensors/sns_reg_version
in the root directory.
---
 hexagonrpcd/hexagonrpcd.1  |  1 +
 hexagonrpcd/rpcd_builder.c | 14 +++++++++++---
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/hexagonrpcd/hexagonrpcd.1 b/hexagonrpcd/hexagonrpcd.1
index 54c77bb..4f8562a 100644
--- a/hexagonrpcd/hexagonrpcd.1
+++ b/hexagonrpcd/hexagonrpcd.1
@@ -43,6 +43,7 @@ acdb                    /vendor/etc/acdbdata
 dsp                     /vendor/dsp
 sensors/config          /vendor/etc/sensors/config
 sensors/registry        /mnt/vendor/persist/sensors/registry/registry
+sensors/sns_reg_version /mnt/vendor/persist/sensors/registry/sns_reg_version
 sensors/sns_reg.conf    /vendor/etc/sensors/sns_reg_config
 socinfo                 /sys/devices/soc0
 .TE
diff --git a/hexagonrpcd/rpcd_builder.c b/hexagonrpcd/rpcd_builder.c
index d374958..bd166e8 100644
--- a/hexagonrpcd/rpcd_builder.c
+++ b/hexagonrpcd/rpcd_builder.c
@@ -31,6 +31,7 @@
 #define SENSORS_CONFIG		"/sensors/config/"
 #define SENSORS_REGISTRY	"/sensors/registry/"
 #define SNS_REG_CONFIG		"/sensors/sns_reg.conf"
+#define SNS_REG_VERSION		"/sensors/sns_reg_version"
 #define SYSFS_SOCINFO		"/socinfo/"
 
 static struct hexagonfs_dirent *hfs_mkdir(const char *name, size_t n_ents, ...)
@@ -111,7 +112,7 @@ static struct hexagonfs_dirent *hfs_map_or_empty(const char *name, const char *p
  */
 struct hexagonfs_dirent *construct_root_dir(const char *prefix, const char *dsp)
 {
-	char *acdbdata, *dsp_libs, *sns_cfg, *sns_reg, *sns_reg_config, *socinfo;
+	char *acdbdata, *dsp_libs, *sns_cfg, *sns_reg, *sns_reg_config, *sns_reg_version, *socinfo;
 	size_t n_prefix;
 	struct hexagonfs_dirent *persist_dir, *vendor_dir;
 
@@ -121,6 +122,7 @@ struct hexagonfs_dirent *construct_root_dir(const char *prefix, const char *dsp)
 	sns_cfg = malloc(n_prefix + strlen(SENSORS_CONFIG) + 1);
 	sns_reg = malloc(n_prefix + strlen(SENSORS_REGISTRY) + 1);
 	sns_reg_config = malloc(n_prefix + strlen(SNS_REG_CONFIG) + 1);
+	sns_reg_version = malloc(n_prefix + strlen(SNS_REG_VERSION) + 1);
 	socinfo = malloc(n_prefix + strlen(SYSFS_SOCINFO) + 1);
 
 	dsp_libs = malloc(n_prefix + strlen(DSP_LIBS) + strlen(dsp) + 1);
@@ -145,6 +147,11 @@ struct hexagonfs_dirent *construct_root_dir(const char *prefix, const char *dsp)
 		strcat(sns_reg_config, SNS_REG_CONFIG);
 	}
 
+	if (sns_reg_version != NULL) {
+		strcpy(sns_reg_version, prefix);
+		strcat(sns_reg_version, SNS_REG_VERSION);
+	}
+
 	if (socinfo != NULL) {
 		strcpy(socinfo, prefix);
 		strcat(socinfo, SYSFS_SOCINFO);
@@ -162,8 +169,9 @@ struct hexagonfs_dirent *construct_root_dir(const char *prefix, const char *dsp)
 	 */
 	persist_dir = hfs_mkdir("persist", 1,
 				hfs_mkdir("sensors", 1,
-					hfs_mkdir("registry", 1,
-						hfs_map("registry", sns_reg)
+					hfs_mkdir("registry", 2,
+						hfs_map("registry", sns_reg),
+						hfs_map("sns_reg_version", sns_reg_version)
 					)
 				)
 		      );
